### PR TITLE
Wikigeo with osm samples

### PR DESCRIPTION
**Pull request by tzufgoogle**
Extended the WikiGeo dataset by adding:
(1) samples based on Wikidata info only
(2) samples based on OSM info only.

Build, run tests working.

### CODE REVIEW

File: cabby/data/extract.py [hunk 1]
```diff
@@ -14,140 +14,197 @@
...
-            if any(dict['text'] == sample_text for dict in geo_data):
-                continue
-            geo_data.append(wikigeo_sample)
-    return geo_data
+from cabby.data import osm_item
+from cabby.geo.map_processing import map_structure
+
+
+def get_wikigeo_data(wikidata_items: Sequence[wdi.WikidataEntity]) -> List:
+  '''Get data from Wikipedia based on Wikidata items"
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:36]: no double quotes here.

File: cabby/data/extract.py [hunk 2]
```diff
@@ -14,140 +14,197 @@
...
+  '''
+  # Get Wikipedia titles.
+  titles = [entity.wikipedia_title for entity in wikidata_items]
+
+  # Get Wikipedia pages.
+  wikipedia_pages = wpq.get_wikipedia_items(titles)
+  wikipedia_items = [wpi.WikipediaEntity.from_api_result(
+    result) for result in wikipedia_pages]
+
+  # # Get Wikipedia titles.
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:52]: may be left from some auto-formatting, but why "# #"? just "#" :)

File: cabby/data/extract.py [hunk 3]
```diff
@@ -14,140 +14,197 @@
...
+  backlinks_pages = wpq.get_backlinks_items_from_wikipedia_titles(
+    wikipedia_titles)
+  backlinks_items = []
+  for list_backlinks in backlinks_pages:
+    backlinks_items.append(
+      [wpi.WikipediaEntity.from_backlinks_api_result(result) for
+        result in list_backlinks])
+
+  # Change backlinks pages to Geodata dataset format.
+  for list_backlinks, original_wikipedia, original_wikidata in \
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:71]: may be just call this backlinks?

File: cabby/data/extract.py [hunk 4]
```diff
@@ -14,140 +14,197 @@
...
+      zip(backlinks_items, wikipedia_items, wikidata_items):
+    for backlink in list_backlinks:
+      wikigeo_sample = wikigeo.WikigeoEntity.from_wiki_items(
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:74]: this is actually the full text right? why sample?

File: cabby/data/extract.py [hunk 5]
```diff
@@ -14,140 +14,197 @@
...
 
 
 def get_data_by_qid(qid: Text) -> Sequence:
-    '''Get data from Wikipedia and Wikidata by region" 
-    Arguments:
-        qid(Text): The qid of the Wikidata to extract items from.
-    Returns:
-        The Wikipedia (text,title) and Wikidata (location) data found.
-    '''
+  '''Get data from Wikipedia and Wikidata by region"
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:85]: replace with .
and everywhere below.

File: cabby/data/extract.py [hunk 6]
```diff
@@ -14,140 +14,197 @@
...
+  # Get Wikidata items by region.
+  wikidata_results = wdq.get_geofenced_wikidata_items(region)
+  wikidata_items = [wdi.WikidataEntity.from_sparql_result(result)
+            for result in wikidata_results]
+
+  return get_wikigeo_data(wikidata_items)
+
+
+def get_data_by_region_expansive(region: Text, path_osm: Text) -> Sequence:
+  '''Get data from Wikipedia and Wikidata by region and add sampes created from
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:117]: typo?

File: cabby/data/extract.py [hunk 7]
```diff
@@ -14,140 +14,197 @@
...
+
+  # Get Wikidata items by region.
+  wikidata_results = wdq.get_geofenced_wikidata_items(region)
+  wikidata_items = [wdi.WikidataEntity.from_sparql_result(result)
+            for result in wikidata_results]
+
+  return get_wikigeo_data(wikidata_items)
+
+
+def get_data_by_region_expansive(region: Text, path_osm: Text) -> Sequence:
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:116]: what does expansive mean?
looking at the code it seems like it is returning the region given the specific region boundaries. may be rename the function as get_data_by_region_boundaries?

File: cabby/data/extract.py [hunk 8]
```diff
@@ -14,140 +14,197 @@
...
+    info_item = wdqi.WikidataEntity.from_sparql_result_info(item)
+    sample = wikigeo.WikigeoEntity.from_wikidata(info_item).sample
+    samples.append(sample)
+
+  logging.info(
+    'Created {} samples with Wikidata additional data.'.format(len(samples)))
+
+  # Add sample from OSM only.
+  poi = map_structure.load_poi(path_osm)
+  poi = poi[poi['s2cellids'].str.len() <= 10]  # Remove large entities.
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:149]: may be define a constant for this?
Also lets count these examples. Are there too many? may be affected by how the OSM entities are named. or rather how certain type of wiki entities are named.

File: cabby/data/extract_wikigeo_contexts_expansive.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+'''Example command line method to extract Wikipedia and Wikidata items and save 
+to file.
+Example:
+$ bazel-bin/cabby/data/extract_wikigeo_contexts_expansive \
```
Comment (sayalikulkarni) [inline cabby/data/extract_wikigeo_contexts_expansive.py:19]: same suggestion about the name as above.

File: cabby/data/extract_wikigeo_contexts_expansive.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+from absl import logging
+from absl import app
+from absl import flags
+import json
+import os
+
+from cabby.data import extract
+
+FLAGS = flags.FLAGS
+FLAGS = flags.FLAGS
```
Comment (sayalikulkarni) [inline cabby/data/extract_wikigeo_contexts_expansive.py:33]: duplicate

File: cabby/data/osm_item.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  """
+  osmid: int = attr.ib()
+  text: Text = attr.ib(init=False)
+  name: Text = attr.ib(init=False)
+  qid: Text = attr.ib(init=False)
+  point: Point = attr.ib()
+  raw: gpd.GeoSeries = attr.ib()
+
+  def __attrs_post_init__(self):
+    # drop_col = ['element_type', 'wheelchair', 'internet_access',
```
Comment (sayalikulkarni) [inline cabby/data/osm_item.py:43]: remove commented code?

File: cabby/data/osm_item.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  def from_osm(cls, row):
+    """Construct an Entity from the Wikidata tags."""
+    return OSMEntity(
+      row['osmid'],
+      row['centroid'],
+      row)
+
+
+def is_english(text):
+  return text.isascii()
```
Comment (sayalikulkarni) [inline cabby/data/osm_item.py:67]: does this drop names like:
https://en.wikipedia.org/wiki/Angoul%C3%AAme

File: cabby/data/osm_item.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      if not is_english(v):
+        continue
+      if 'colour' in k and v[0]=='#':
+        v = get_colour_name(v)
+      k = k.replace("_", " ")
+      v = v.replace("_", " ")
+      if k in ['building','historic']:  
+        text += connect + k
+      else:
+        text += connect + v
```
Comment (sayalikulkarni) [inline cabby/data/osm_item.py:84]: you can do:
k_text = []
v_text = []

if  ... : 
  k_text.append(k)
else:
  v_text.append(v)

and then at the end - 
" and ".join(v_text, k_text)

confirm that this works ok for the sequence you need before updating.

File: cabby/data/wikidata/info_item.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+  `sample` is created from a concatenation of the entity's tags.
+
+  """
+  url: Text = attr.ib()
+  title: Text = attr.ib()
+  location: Point = attr.ib()
+  instance: Text = attr.ib()
+  qid: Text = attr.ib(init=False)
+  tags: Dict = attr.ib()
+  sample: Text = attr.ib(init=False)
```
Comment (sayalikulkarni) [inline cabby/data/wikidata/info_item.py:48]: relates to one of the comments above. the attribute can be named tags instead of sample?

**Review by sayalikulkarni — CHANGES_REQUESTED** [review]
This too seems like a very large PR :) 
it is easier to test, review and submit smaller code pieces rather than one large PR. May be try to split more next time?

**Review by tzufgoogle — COMMENTED** [review]
I didn't change the function name - it expands over another function with the addition of 2 new types of samples.
resolved conflicts.
build, run, test working.

File: cabby/data/extract.py (hunk 5 above)
Comment (sayalikulkarni) [reply cabby/data/extract.py:85]: Are the changes merged? Seems like I still see it as a "

File: cabby/data/extract.py (hunk 7 above)
Comment (sayalikulkarni) [reply cabby/data/extract.py:116]: may be rename as get_data_by_region_with_osm?
'expansive' is not very evident

File: cabby/data/extract.py [hunk 15]
```diff
@@ -14,140 +14,198 @@
...
+      
+  logging.info(
+    'Created {} samples with OSM additional data.'.format(len(samples)))
+  return samples
+
+
+def split_dataset(
     dataset: Sequence, percentage_train: float, percentage_dev: float):
-    '''Splits the dataset into train-set, dev-set, test-set according to the 
-    ref_qid"
```
Comment (sayalikulkarni) [inline cabby/data/extract.py:111]: pending fix?

File: cabby/data/extract_wikigeo_contexts_expansive.py (hunk 10 above)
Comment (sayalikulkarni) [reply cabby/data/extract_wikigeo_contexts_expansive.py:33]: still pending?

**Review by sayalikulkarni — CHANGES_REQUESTED** [review]
There still seem to be some pending changes? Is there a pending commit to be pushed?

**Review by tzufgoogle — COMMENTED** [review]
Changed file name

**Pull request merged**